Base submittable item API

aschroed opened the pull request:
Base framework for the submittable item API.
There may be some outstanding questions as to appropriate defaults - currently an empty dict is returned if no criteria are provided for identifying a submittable schema, which I think is what we want.  
Options for identifying a schema as submittable include an explicit list of schema names or a specific property to look for in a schema - this is set at the project level.

There is some hard coding of schema features such as 'required', 'oneOf' and 'required_dependent' which are kind of built ins of the structure of our json schemas but it may be desirable to make this more configurable or have constants for these?

I haven't had time and not sure exactly the best way to set up tests and mocks for this so help on that would be appreciated.  There is currently very minimal testing.

The project specific extension of this for the smaht-portal is also implemented but was built on top of the submittable data model work that is not in main yet so haven't made a PR for that.

willronchetti commented on snovault/schema_views.py:
```diff
@@ -116,3 +120,158 @@ def schemas(context, request):
         schemas[name] = _annotated_schema(type_info, request)
     return schemas
+
+def _get_propnames_from_oneof(oneof_info):
+    propnames = []
+    for oneof in oneof_info:
+        # is a list but examples so far have only one member?
+        propnames.extend(oneof.get('required', []))
+    return propnames
```
This function name is a bit confusing to me because it doesn't really indicate you're doing an operation specifically on the required props... Some clarification would be helpful I think

willronchetti commented on snovault/schema_views.py:
```diff
@@ -116,3 +120,158 @@ def schemas(context, request):
+    return propnames
+
+
+def _has_attr(propinfo, include_attrs):
+    for aname, avalues in include_attrs.items():
+        if aname in propinfo:
+            if propinfo.get(aname) in avalues:
+                return True
+    return False
```
Again slightly confused by function name here, suggests a fairly simple/general operation but seems more complex than that... docstrings would help even for "hidden" functions

willronchetti commented on snovault/schema_views.py:
```diff
@@ -116,3 +120,158 @@ def schemas(context, request):
+    # eg. submmitter_id
+    key_prop = app_project().get_prop_for_submittable_items()
+
+    # explicit list of propnames to exclude eg. 'last_modified'
+    excluded_props = app_project().get_properties_for_exclusion()
+    # explicit list of propnames to include no current examples
+    included_props = app_project().get_properties_for_inclusion()
```
Can we simplify and just have the "include" list?

aschroed replied:
General reply to your simplification question - it we were only going to have either included or excluded then excluded is the way we would want to go, otherwise you would need to list all the properties that you want to include, which is most of them.  The idea here (and the reason for the 'odd' logic that you ask about later on in the _get_submittable_schema function) is to allow you to flexibly specify those properties or attributes you definitely want included even if they contain an attribute that would normally exclude them.  Hence inclusion trumps exclusion in the if: elif: elif: bit below.

And having the ability to exclude specific properties or properties that have specific attributes is key to only providing submittable properties.  For example, consortia and submission_center don't have permission: restricted_fields (and as I write this I am realizing I may not be correct in my assumption) but I figured that these properties would not be included in the sheet metadata but could be inferred from the user (for Submit4DN we infer and ask if there are more than one award or lab associated with a submitter).  In any case my assumption was that these should be excluded from the submittable properties so could be listed in the smaht_portal app_project.get_properties_for_exclusion list.  And we want to exclude any property that has the attribute permission: restricted_field.

A completely made up unlikely scenario is what if you wanted a users to be able to submit 'status' - it has the permission: restricted_field so normally would be filtered out but if you added 'status' into the included properties list then it would be included in the submittable schema - not that you'd want it there as the submission would fail.  

So yes it may be more complex than needed - I was just trying to cover all the bases.  I'll have a look at the submittable branch of smaht to determine if inclusion is definitely overkill and if so remove it.

willronchetti replied:
It seems like then instead of just having an "allow" list you could get away with just a "deny" list per type in addition to a global list? My point is simply having one can completely represent every scenario you want I think... If most properties are submittable then a "deny" list makes sense, especially since a lot of those will be shared across collections? Maybe I'm not understanding the situation you're thinking of where both are needed, but it's not obvious to me why both are necessary.

willronchetti commented on snovault/schema_views.py:
```diff
@@ -116,3 +120,158 @@ def schemas(context, request):
+    # because there could be different values of an attribute this a dictionary
+    # keyed by attribute name with a list of values
+    # eg. {'permission':['restricted_field']}
+    exclude_attrs = app_project().get_attributes_for_exclusion()
+    # explicit attrs to allow inclusion in submit props
+    # currently not used but will have same structure as above
+    include_attrs = app_project().get_attributes_for_inclusion()
```
Same deal here - why need an exclude list when there is an include list?

willronchetti commented on snovault/schema_views.py:
```diff
@@ -116,3 +120,158 @@ def schemas(context, request):
+        # an explicit schema list takes preference over a identifying field to determine submittableness    
+        if not schema_id:
+            return {}
+        schema_name = schema_id.replace('/profiles/', '').replace('.json', '')
```
Might want a helper for this

willronchetti commented on snovault/schema_views.py:
```diff
@@ -116,3 +120,158 @@ def schemas(context, request):
+            return {}
+    else:
+        # if neither present in first case say it's not submittable - do we want a different default behavior?
+        return{}
```
Space

willronchetti commented on snovault/tests/test_schemas.py:
```diff
@@ -31,3 +31,15 @@ def test_schemas_etag(testapp):
+    res = testapp.get(test_uri, status=200)
+    assert not res.json
+
+def test_submittables(testapp, registry):
```
I would recommend adding a unit test that includes a submittable schema here, just use a simple dummy definition in the unit tests and mock the relevant app_project calls

drio18 replied:
Yeah, having tests on dummy schemas would be great here, and breaking up that large function into smaller ones would simplify unit testing greatly as well as later updates.

willronchetti commented on snovault/schema_views.py:
```diff
@@ -116,3 +120,158 @@ def schemas(context, request):
+    for propname, propinfo in schema_props.items():
+        emb_obj = None
+        # determine if prop should be submittable - inclusion trumps exclusion
+        if propname in included_props:  # independent of any other attribute
+            pass
+        elif propname in excluded_props:  # explicity excluded by name
+            continue
+        elif _has_attr(propinfo, include_attrs):  # check for explicit attr that indicate inclusion
+            pass
+        elif _has_attr(propinfo, exclude_attrs):
+            continue
```
This pattern is a little odd and can be simplified no? Especially if taking prior suggestion for simplification ie:
```
if propname not in included_props:
    continue
elif not _has_attr(propinfo, include_attrs):
    continue
```
and so on?

willronchetti reviewed (approved):
Some comments I think merit addressing and should probably include at least one unit test with a simple collection

drio18 commented on snovault/schema_views.py:
```diff
@@ -116,3 +120,162 @@ def schemas(context, request):
+def submittable(context, request):
+    # import pdb; pdb.set_trace()
+    type_name = request.matchdict['type_name']
+    schema = load_schema(f"schemas/{type_name}.json")
```
Consider gathering the schema via `_annotated_schema` in line 26 above rather than via `load_schema`. I think the former should have the schema cached and ready to roll, whereas the latter will repeat a lot of this work (loading the file, resolving mixins/merge references, etc.). Pretty sure this is the case anyway.

aschroed replied:
I thought I remember @willronchetti suggesting to use the schema_utils load_schema function.  I certainly could make a subrequest to the schema function - otherwise not sure exactly how to get the type info needed for the annotated schema function and not clear on what might be cached that would make this more efficient?  If there is a better way to do this I'm game - just not sure the best way to do so.

drio18 replied:
If you don't want to go the subrequest route about which adds further schema annotations, some of which may be useful to include here, you can alternatively grab the schemas from the registry as done in the `schemas` definition above: https://github.com/4dn-dcic/snovault/blob/master/snovault/schema_views.py#L111-L118. 

This also provides an easy route for excluding abstract collections, which should probably be the default here but I don't believe is currently handled.

aschroed replied:
Will look at this some more.

drio18 commented on snovault/schema_views.py:
```diff
@@ -116,3 +120,162 @@ def schemas(context, request):
+    if not schema_props:
+        return {}
+
+    # first determine if the schema is submittable
+    is_submittable = False
+    if is_embedded_obj:  # infrequent case from recursive call
+        is_submittable = True
+
+    if item_list and not is_submittable:
+        # an explicit item list takes preference over a identifying field to determine submittableness
+        # if item is in the list it is considered submittable    
+        if schema_id:
+            item_name = _get_item_name_from_schema_id(schema_id)
+            if item_name in item_list:
+                is_submittable = True
+    
+    if key_prop:
+        # a property that if in the schema identifies the schema for a submittable item
+        prop_names = schema_props.keys()
+        if key_prop in prop_names:
+            is_submittable = True
```
Consider breaking this large function into multiple small chunks to organize the high-level logic a bit. Some of these chunks are easy enough to drop into their own functions, separating all of the is submittable part from the collection of the submittable properties.

drio18 commented on snovault/schema_views.py:
```diff
@@ -116,3 +120,162 @@ def schemas(context, request):
+    required_props = schema.get('required', [])
+    
+    oneof_props = _get_required_propnames_from_oneof(schema)
+    req_deps = schema.get('dependent_required', {})
```
This should actually be "dependentRequired", which has been fixed in my branch: https://json-schema.org/understanding-json-schema/reference/conditionals#dependentRequired

drio18 commented on snovault/project/schema_views.py:
```diff
@@ -0,0 +1,18 @@
+class SnovaultProjectSchemaViews:
+
+    def get_submittable_item_names(self):
+        # optional project dependent list of item names in snake case format to include as submittable
```
Consider putting into docstring rather than comment form.

drio18 reviewed (commented):
Overall, looks good and flexible enough for the ask. Left a few comments for consideration.

aschroed commented:
@drio18 and @willronchetti I think this is ready for another look.  Refactored code and added several tests (thanks for the help Will). 

Outstanding questions:
1. Will had left a comment about possibly using Title instead of $id in a particular test, however, the functions currently use and somewhat depend on a schema having a $id value corresponding to the snake case profiles 'path' to the schema which is used to get the snake case schema name that is used for loading.  Title values could be used and converted as needed but I was going with the $id being more reliable but maybe for no good reason.

2. Doug had wondered about using the _annotated_schema function to get all the schemas which may utilize caching but there are other things going on in that and the calling functions that I was not sure made this a good or easy to implement idea.  I'd appreciate Will's thoughts on that.

drio18 commented on snovault/tests/test_schemas.py:
```diff
@@ -31,3 +47,349 @@ def test_schemas_etag(testapp):
+        assert ans is False
+
+
+def test_submittable_no_app_info(testapp):
+    """ without providing item name or key prop we expect
+        a 200 response but no json
+    """
+    test_uri = '/can-submit/access_key.json'
+    res = testapp.get(test_uri, status=200)
+    assert not res.json
```
Perhaps this should raise an HTTP error rather than indicate success? On a similar note, should probably raise an HTTP error if someone requests an item type that does not have a schema, which I think would currently result in the same response here that doesn't really provide information. Consider adding a test for this situation to state the behavior more explicitly.

aschroed replied:
I will look into this in the context of raising exceptions in certain situations.  Thanks for the suggestion.

drio18 commented on snovault/schema_views.py:
```diff
@@ -11,11 +11,15 @@
 def includeme(config):
     config.add_route('schemas', '/profiles/')
     config.add_route('schema', '/profiles/{type_name}.json')
+    config.add_route('submittables', '/can-submit/')
+    config.add_route('submittable', '/can-submit/{type_name}.json')
```
Consider changing the route from `can-submit` to `submission-schemas` or something along those lines; `can-submit` suggests a more boolean operation to me. Admittedly a subjective comment and entirely your call here.

aschroed replied:
agreed - done

drio18 reviewed (approved):
Dropping a few comments for consideration but nothing blocking. Tests look good and think this is a great foundation to build from.

aschroed commented on snovault/schema_views.py:
```diff
@@ -116,3 +119,186 @@ def schemas(context, request):
+    for propname, propinfo in props.items():
+        if propname in required_props:
+            propinfo['is_required'] = True
+        if propname in oneof_props:  # this a bit wonky (assumes only 2 items in oneof_props)
+            propinfo['required_if_not'] = [p for p in oneof_props if p != propname][0]
```
Yep makes sense.

aschroed commented on snovault/schema_views.py:
```diff
@@ -116,3 +119,186 @@ def schemas(context, request):
+    submittable props based on the schema info
+    """
+    required_props = schema.get('required', [])
+    oneof_props = _get_required_propnames_from_oneof(schema)
```
added

aschroed commented:
I've incorporated most of Doug's suggestions and confirmed that things appear to work as expected when this branch is used in the corresponding branch of the smaht-portal (at least at the level of basic tests).  Looking at the new schema_utils a couple things might be adjusted and I'm currently looking into that but not sure if it should hold this up.

willronchetti commented on snovault/schema_views.py:
```diff
@@ -116,3 +126,201 @@ def schemas(context, request):
+    propnames = []
+    keyword_info = schema.get(keyword, [])
+    for info in keyword_info:
+        # is a list but examples so far have only one member?
```
Just for future flexibility I'm sure

willronchetti commented on snovault/schema_views.py:
```diff
@@ -116,3 +126,201 @@ def schemas(context, request):
+    return False
+
+
+def _get_item_name_from_schema_id(schema_id):
+    """ this assumes a validly formatted $id value from a schema"""
+    return schema_id.replace('/profiles/', '').replace('.json', '')
```
You might want to `.lower()` as well to further control for incompatibilities that may crop up, but not sure it's really necessary

willronchetti commented on snovault/schema_views.py:
```diff
@@ -116,3 +126,201 @@ def schemas(context, request):
+    required_props = get_required(schema)
+    oneof_props = _get_conditionally_required_propnames(schema, sc.ONE_OF)
+    anyof_props = _get_conditionally_required_propnames(schema, sc.ANY_OF)
+    req_deps = schema.get('dependentRequired', {})
+
+    for propname, propinfo in props.items():
+        if propname in required_props:
+            propinfo['is_required'] = True
+        if propname in oneof_props:
+            lprops = [p for p in oneof_props if p != propname]
+            propinfo.setdefault('required_if_not_one_of', []).extend(lprops)
+            propinfo['prohibited_if_one_of'] = lprops
+        if propname in anyof_props:
+            lprops = [p for p in anyof_props if p != propname]
+            propinfo.setdefault('required_if_not_one_of', []).extend(lprops)
+        if propname in req_deps:
+            propinfo['also_requires'] = req_deps[propname]
```
You might consider making some of the dictionary keys here constants, though I'd especially consider those that are tied to `jsonschema` itself (and not structure you're defining)

willronchetti reviewed (approved):
Recent changes look good, just some small comments

drio18 reviewed (approved):
Looks good.
